sdk/lib/opentelemetry/sdk/trace/span.rb

```diff
@@ -7,9 +7,236 @@
 module OpenTelemetry
   module SDK
     module Trace
+      # Implementation of {OpenTelemetry::Trace::Span} that records trace events.
+      #
+      # rubocop:disable Metrics/ClassLength
       class Span < OpenTelemetry::Trace::Span
-        # TODO
+        # TODO: does this need synchronization? I don't think so...
+        attr_reader :name
+
+        # Return the flag whether this span is recording events
+        #
+        # @return [Boolean] true if this Span is active and recording information
+        #   like events with the #add_event operation and attributes using
+        #   #set_attribute.
+        def recording_events?
+          true
+        end
+
+        # Set attribute
+        #
+        # Note that the OpenTelemetry project
+        # {https://github.com/open-telemetry/opentelemetry-specification/blob/master/semantic-conventions.md
+        # documents} certain "standard attributes" that have prescribed semantic
+        # meanings.
+        #
+        # @param [String] key
+        # @param [String, Boolean, Numeric] value
+        #
+        # @return [self] returns itself
+        def set_attribute(key, value)
+          super
+          @mutex.synchronize do
+            if @ended
+              logger.debug('Calling set_attribute on an ended Span.')
+            else
+              @attributes ||= {}
+              @attributes[key] = value
+              trace_config.trim_attributes(@attributes, :max_attributes_count)
+              @total_recorded_attributes += 1
+            end
+          end
+          self
+        end
+
+        # Add an Event to Span
+        #
+        # Note that the OpenTelemetry project
+        # {https://github.com/open-telemetry/opentelemetry-specification/blob/master/semantic-conventions.md
+        # documents} certain "standard event names and keys" which have
+        # prescribed semantic meanings.
+        #
+        # @param [String, Callable] name_or_event_formatter The name of the event
+        #   or an EventFormatter, a lazily evaluated callable that returns an
+        #   Event instance.
+        # @param [optional Hash<String, Object>] attrs One or more key:value pairs, where
+        #   the keys must be strings and the values may be string, boolean or
+        #   numeric type. This argument should only be used when passing in a
+        #   name, not an EventFormatter.
+        # @param [Time] timestamp optional timestamp for the event.
+        #
+        # @return [self] returns itself
+        def add_event(name_or_event_formatter, attrs = nil, timestamp: nil)
+          super
+          timed_event =
+            if name_or_event_formatter.instance_of?(String)
+              TimedEvent.new(name_or_event_formatter, attrs, timestamp || Time.now)
+            else
+              name_or_event_formatter.call
```

Comment:
in this case should you pass the attrs and timestamp to the .call ?

Reply:
No. The attrs and timestamp are only supplied for "eagerly evaluated" Events, not for lazily-evaluated ones.